drivers/platform/surface/surface_fan.c

```diff
@@ -0,0 +1,287 @@
+			// No known way to retrieve the current setpoint.
+			return -1;
+		default:
+			break;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return -1;
+}
+
+
+static int surface_fan_hwmon_write(struct device *dev,
+					enum hwmon_sensor_types type,
+					u32 attr, int channel, long val)
+{
+	__le16 value;
+	struct fan_data *data = dev_get_drvdata(dev);
+
+	switch (type) {
+	case hwmon_fan:
+		switch (attr) {
+		case hwmon_fan_target:
+			value = cpu_to_le16(clamp(val, 0l, (1l << 16)));
+			return __ssam_fan_set(data->ctrl, &value);
+		default:
+			break;
+		}
+		break;
+	default:
+		break;
+	}
+	return -1;
+}
+
+static const struct hwmon_channel_info * const surface_fan_info[] = {
+	HWMON_CHANNEL_INFO(fan,
+				HWMON_F_INPUT |
+				HWMON_F_MAX |
+				HWMON_F_MIN |
+				HWMON_F_TARGET
+				),
+	NULL
+};
+
+static const struct hwmon_ops surface_fan_hwmon_ops = {
+	.is_visible = surface_fan_hwmon_is_visible,
+	.read = surface_fan_hwmon_read,
+	.write = surface_fan_hwmon_write,
+};
+
+static const struct hwmon_chip_info surface_fan_chip_info = {
+	.ops = &surface_fan_hwmon_ops,
+	.info = surface_fan_info,
+};
+
+static int surface_fan_probe(struct platform_device *pdev)
+{
+	struct acpi_device *acpi_fan = ACPI_COMPANION(&pdev->dev);
+
+	struct ssam_controller *ctrl;
+	struct fan_data *data;
+	struct thermal_cooling_device *cdev;
+	struct device *hdev;
+	__le16 value;
+	int status;
+
+	ctrl = ssam_client_bind(&pdev->dev);
+	if (IS_ERR(ctrl))
+		return PTR_ERR(ctrl) == -ENODEV ? -EPROBE_DEFER : PTR_ERR(ctrl);
+
+	// Probe the fan to confirm we actually have it by retrieving the 
+	// speed.
+	status = __ssam_fan_get(ctrl, &value);
+	if (status) {
+		return -ENODEV;
+	}
+
+	data = devm_kzalloc(&pdev->dev, sizeof(*data), GFP_KERNEL);
+	if (!data)
+		return -ENOMEM;
+
+	cdev = thermal_cooling_device_register("Fan",
+					data, &surface_fan_cooling_ops);
+	if (IS_ERR(cdev))
+		return PTR_ERR(cdev);
+
+
+	hdev = devm_hwmon_device_register_with_info(&pdev->dev, "fan", data,
+							&surface_fan_chip_info,
+							NULL);
+	if (IS_ERR(hdev)) {
+		return PTR_ERR(hdev);
+	}
+
+	data->dev = &pdev->dev;
+	data->ctrl = ctrl;
+	data->acpi_fan = acpi_fan;
+
+	data->cdev = cdev;
+	data->hdev = hdev;
+
+	acpi_fan->driver_data = data;
+	platform_set_drvdata(pdev, data);
+
+	return 0;
+}
+
+static int surface_fan_remove(struct platform_device *pdev)
+{
+	struct fan_data *d = platform_get_drvdata(pdev);
+	thermal_cooling_device_unregister(d->cdev);
+	return 0;
+}
+
+
+static struct platform_driver surface_fan = {
+	.probe = surface_fan_probe,
+	.remove = surface_fan_remove,
+	.driver = {
+		.name = "surface_fan",
+		.acpi_match_table = surface_fan_match,
+		.probe_type = PROBE_PREFER_ASYNCHRONOUS,
+	},
+};
+
+static int __init surface_fan_init(void)
+{
+	return platform_driver_register(&surface_fan);
+}
+module_init(surface_fan_init);
+
+static void __exit surface_fan_exit(void)
+{
+	platform_driver_unregister(&surface_fan);
+}
+module_exit(surface_fan_exit);
```

Comment:
This all can be replaced with `module_platform_driver(surface_fan);`, which does exactly the same.